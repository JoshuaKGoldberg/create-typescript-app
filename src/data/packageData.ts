import { createRequire } from "node:module";

const require = createRequire(import.meta.url);

export const packageData =
	// Importing from above src/ would expand the TS build rootDir
	require("../../package.json") as typeof import("../../package.json");

export function getPackageDependencies(...names: string[]) {
	return Object.fromEntries(
		names.map((name) => {
			return [name, getPackageDependency(name)];
		}),
	);
}

export function getPackageDependency(name: string) {
	const version =
		getPackageInner("devDependencies", name) ??
		getPackageInner("dependencies", name);

	if (!version) {
		throw new Error(
			`'${name} is neither in package.json's dependencies nor its devDependencies.`,
		);
	}

	return version;
}

function getPackageInner(
	key: "dependencies" | "devDependencies",
	name: string,
) {
	const inner = packageData[key];

	return inner[name as keyof typeof inner] as string | undefined;
}
