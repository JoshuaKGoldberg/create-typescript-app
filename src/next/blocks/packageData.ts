import { createRequire } from "node:module";

const require = createRequire(import.meta.url);

const packageData =
	// Importing from above src/ would expand the TS build rootDir

	require("../../../package.json") as typeof import("../../../package.json");

const getPackageInner = (
	key: "dependencies" | "devDependencies",
	name: string,
) => {
	const inner = packageData[key];

	return inner[name as keyof typeof inner] as string | undefined;
};

export const getPackageDependencies = (...names: string[]) =>
	Object.fromEntries(
		names.map((name) => {
			const version =
				getPackageInner("devDependencies", name) ??
				getPackageInner("dependencies", name);

			if (!version) {
				throw new Error(
					`'${name} is neither in package.json's dependencies nor its devDependencies.`,
				);
			}

			return [name, version];
		}),
	);
