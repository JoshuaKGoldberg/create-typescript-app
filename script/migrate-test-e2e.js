import chalk from "chalk";
import { $, execaCommand } from "execa";
import fs from "node:fs/promises";
import { assert, describe, expect, test } from "vitest";

import packageData from "../package.json" assert { type: "json" };

const filesExpectedToBeChanged = [
	"README.md",
	"knip.jsonc",
	"package.json",
	".eslintignore",
	".eslintrc.cjs",
	".github/workflows/test.yml",
	".gitignore",
	".prettierignore",
	"cspell.json",
];

const filesThatMightBeChanged = new Set([
	...filesExpectedToBeChanged,
	"script/__snapshots__/migrate-test-e2e.js.snap",
]);

const {
	author: { email: emailNpm, name: authorName },
	description,
	name: repository,
} = packageData;
const emailGithub = "github@joshuakgoldberg.com";
const bin = "./bin/index.js";
const guide =
	"https://www.joshuakgoldberg.com/blog/contributing-to-a-create-typescript-app-repository";
const guideTitle = "Contributing to a create-typescript-app Repository";
const logo = "./create-typescript-app.png";
const logoAlt = `Project logo: the TypeScript blue square with rounded corners, but a plus sign instead of 'TS'`;
const owner = "JoshuaKGoldberg";
const title = "Create TypeScript App";

const originalDevelopment = (
	await fs.readFile(".github/DEVELOPMENT.md")
).toString();

const originalReadme = (await fs.readFile("README.md")).toString();

const originalSnapshots = (
	await fs.readFile("script/__snapshots__/migrate-test-e2e.js.snap")
).toString();

await $({
	stdio: "inherit",
})`c8 -o ./coverage -r html -r lcov --src src node ${bin} --base everything --author ${authorName} --mode migrate --bin ${bin} --description ${description} --email-github ${emailGithub} --email-npm ${emailNpm} --guide ${guide} --guide-title ${guideTitle} --logo ${logo} --logo-alt ${logoAlt} --owner ${owner} --title ${title} --repository ${repository} --skip-all-contributors-api --skip-github-api --skip-install`;

// All Contributors seems to not be using Prettier to format files...
await fs.writeFile(
	".all-contributorsrc",
	JSON.stringify(
		JSON.parse((await fs.readFile(".all-contributorsrc")).toString()),
		null,
		2,
	) + "\n",
);

// The development setup scripts docs can be ignored from snapshots.
// We manually add them back after hydration to clear them from Git diffs.
await fs.appendFile(
	".github/DEVELOPMENT.md",
	"\n" +
		originalDevelopment.slice(originalDevelopment.indexOf("## Setup Scripts")),
);

// Ignore changes to the README.md all-contributor count and contributors table...
const updatedReadme = (await fs.readFile("README.md")).toString();
await fs.writeFile(
	"README.md",
	[
		updatedReadme.slice(0, updatedReadme.indexOf("## Contributors")) +
			originalReadme.slice(
				originalReadme.indexOf("## Contributors"),
				originalReadme.indexOf("<!-- markdownlint-restore -->"),
			),
		updatedReadme.slice(updatedReadme.indexOf("<!-- markdownlint-restore -->")),
	]
		.join("")
		.replaceAll(
			/All Contributors: \d+/g,
			originalReadme.match(/All Contributors: \d+/)[0],
		)
		.replaceAll(
			/all_contributors-\d+/g,
			originalReadme.match(/all_contributors-\d+/)[0],
		),
);

// ...and even to the snapshot file, so diffs don't mind it.
await fs.writeFile(
	"script/__snapshots__/migrate-test-e2e.js.snap",
	originalSnapshots
		.replaceAll(
			/All Contributors: \d+/g,
			originalReadme.match(/All Contributors: \d+/)[0],
		)
		.replaceAll(
			/all_contributors-\d+/g,
			originalReadme.match(/all_contributors-\d+/)[0],
		),
);

describe("expected file changes", () => {
	test.each(filesExpectedToBeChanged)("%s", async (file) => {
		const { stdout } = await execaCommand(`git diff HEAD -- ${file}`);
		const contentsAfterGitMarkers = stdout
			.split("\n")
			.slice(2)
			.join("\n")
			.replaceAll(/@@ -\d+,\d+ \+\d+,\d+ @@/g, "@@ ... @@");

		assert(
			stdout,
			`Looks like there were no changes to ${file} from migration?`,
		);

		expect(contentsAfterGitMarkers).toMatchSnapshot();
	});
});

// eslint-disable-next-line vitest/expect-expect
test("unexpected file changes", async () => {
	const { stdout: gitStatus } = await $`git status`;
	console.log(`Stdout from running \`git status\`:\n${gitStatus}`);

	const indexOfUnstagedFilesMessage = gitStatus.indexOf(
		"Changes not staged for commit:",
	);
	assert(
		indexOfUnstagedFilesMessage !== -1,
		`Looks like migrate didn't cause any file changes? That's ...probably incorrect? ðŸ˜¬`,
	);

	const unstagedModifiedFiles = gitStatus
		.slice(indexOfUnstagedFilesMessage)
		.match(/modified: {3}(\S+)\n/g)
		.map((match) => match.split(/\s+/g)[1])
		.filter((filePath) => !filesThatMightBeChanged.has(filePath));

	console.log("Unexpected modified files are:", unstagedModifiedFiles);

	if (unstagedModifiedFiles.length) {
		const gitDiffCommand = `git diff HEAD -- ${unstagedModifiedFiles.join(
			" ",
		)}`;
		console.log(
			`Stdout from running \`${gitDiffCommand}\`:\n${
				(await execaCommand(gitDiffCommand)).stdout
			}`,
		);
		throw new Error(
			[
				"",
				"Oh no! Running the migrate script modified some files:",
				...unstagedModifiedFiles.map((filePath) => ` - ${filePath}`),
				"",
				"That likely indicates changes made to the repository without",
				"corresponding updates to templates in src/.",
				"",
				"Please search for those file(s)' name(s) under src/migrate for",
				"the corresponding template and update those as well.",
			]
				.map((line) => chalk.red(line))
				.join("\n"),
		);
	}
});
