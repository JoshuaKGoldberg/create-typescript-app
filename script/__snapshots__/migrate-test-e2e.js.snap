// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expected file changes > .eslintignore 1`] = `
"--- a/.eslintignore
+++ b/.eslintignore
@@ ... @@
 !.*
-coverage*
+coverage
 lib
 node_modules
 pnpm-lock.yaml"
`;

exports[`expected file changes > .eslintrc.cjs 1`] = `
"--- a/.eslintrc.cjs
+++ b/.eslintrc.cjs
@@ ... @@
-/*
-ðŸ‘‹ Hi! This ESLint configuration contains a lot more stuff than many repos'!
-You can read from it to see all sorts of linting goodness, but don't worry -
-it's not something you need to exhaustively understand immediately. ðŸ’™
-
-If you're interested in learning more, see the 'getting started' docs on:
-- ESLint: https://eslint.org
-- typescript-eslint: https://typescript-eslint.io
-*/
-
 /** @type {import("@types/eslint").Linter.Config} */
 module.exports = {
 	env: {
@@ ... @@ module.exports = {
 			rules: {
 				// These off-by-default rules work well for this repo and we like them on.
 				"deprecation/deprecation": "error",
-
-				// These more-strict-by-default rules don't work well for this repo and we like them less strict.
-				"@typescript-eslint/no-unnecessary-condition": [
-					"error",
-					{
-						allowConstantLoopConditions: true,
-					},
-				],
-				"@typescript-eslint/prefer-nullish-coalescing": [
-					"error",
-					{ ignorePrimitives: true },
-				],
 			},
 		},
 		{"
`;

exports[`expected file changes > .github/workflows/test.yml 1`] = `
"--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ ... @@ jobs:
       - run: pnpm run test --coverage
       - name: Codecov
         uses: codecov/codecov-action@v3
-        with:
-          flags: unit
-      - if: always()
-        name: Archive code coverage results
-        uses: actions/upload-artifact@v4
-        with:
-          name: code-coverage-report
-          path: coverage
 
 name: Test
 "
`;

exports[`expected file changes > .gitignore 1`] = `
"--- a/.gitignore
+++ b/.gitignore
@@ ... @@
-coverage*/
+coverage/
 lib/
 node_modules/"
`;

exports[`expected file changes > .prettierignore 1`] = `
"--- a/.prettierignore
+++ b/.prettierignore
@@ ... @@
 .all-contributorsrc
-coverage*/
+coverage/
 lib/
 pnpm-lock.yaml"
`;

exports[`expected file changes > README.md 1`] = `
"--- a/README.md
+++ b/README.md
@@ ... @@
 
 <img align="right" alt="Project logo: the TypeScript blue square with rounded corners, but a plus sign instead of 'TS'" src="./create-typescript-app.png">
 
-\`create-typescript-app\` is a one-stop-shop solution to set up a new or existing repository with the latest and greatest TypeScript tooling.
-It includes options not just for building and testing but also GitHub repository templates, contributor recognition, automated release management, and more.
-
-## Getting Started
-
-First make sure you have the following installed:
-
-- [GitHub CLI](https://cli.github.com) _(you'll need to be logged in)_
-- [Node.js](https://nodejs.org)
-- [pnpm](https://pnpm.io)
-
-Then in an existing repository or in your directory where you'd like to make a new repository:
-
-\`\`\`shell
-npx create-typescript-app
-\`\`\`
-
-That setup script will walk you through using the template.
-You can read more about the supported setup modes in their docs pages:
-
-- [**Creating from the terminal**](./docs/Creation.md): creating a new repository locally on the command-line _(recommended)_
-- [**Initializing from the template**](./docs/Initialization.md): creating a new repository with the [_Use this template_](https://github.com/JoshuaKGoldberg/create-typescript-app/generate) button on GitHub
-- [**Migrating an existing repository**](./docs/Migration.md): adding this template's tooling on top of an existing repository
-
-## Documentation
-
-You can read more about \`create-typescript-app\` and the tooling it supports:
-
-1. [**Tooling**](./docs/Tooling.md): a breakdown of all the pieces this template can set up.
-2. [**Options**](./docs/Options.md): granular options to customize how the template is run.
-3. [**FAQs**](./docs/FAQs.md): frequently asked questions
-
-> [!NOTE]  
-> This template is early stage, opinionated, and not endorsed by the TypeScript team.
-> It can be configured to set up a _lot_ of tooling out of the box.
-> If you don't want to use any particular tool, you can always remove it manually.
-
 ## Development
 
 See [\`.github/CONTRIBUTING.md\`](./.github/CONTRIBUTING.md), then [\`.github/DEVELOPMENT.md\`](./.github/DEVELOPMENT.md).
@@ ... @@ Thanks! ðŸ’–
 
 <!-- ALL-CONTRIBUTORS-LIST:END -->
 <!-- spellchecker: enable -->
+
+<!-- You can remove this notice if you don't want it ðŸ™‚ no worries! -->
+
+> ðŸ’™ This package was templated with [\`create-typescript-app\`](https://github.com/JoshuaKGoldberg/create-typescript-app)."
`;

exports[`expected file changes > cspell.json 1`] = `
"--- a/cspell.json
+++ b/cspell.json
@@ ... @@
 {
 	"dictionaries": ["typescript"],
 	"ignorePaths": [
-		"./coverage*",
 		".github",
 		"CHANGELOG.md",
+		"coverage",
 		"lib",
 		"node_modules",
-		"pnpm-lock.yaml",
-		"./script/__snapshots__"
+		"pnpm-lock.yaml"
 	],
 	"words": [
-		"allcontributors",
-		"apexskier",
-		"arethetypeswrong",
 		"Codecov",
 		"codespace",
 		"commitlint",
 		"contributorsrc",
 		"conventionalcommits",
-		"execa",
-		"infile",
-		"joshuakgoldberg",
 		"knip",
 		"lcov",
 		"markdownlintignore",
-		"mtfoley",
-		"npmignore",
-		"npmjs",
 		"npmpackagejsonlintrc",
 		"outro",
 		"packagejson",
 		"quickstart",
-		"tada",
 		"tsup",
-		"Unstaged",
-		"vitest",
 		"wontfix"
 	]
 }"
`;

exports[`expected file changes > knip.jsonc 1`] = `
"--- a/knip.jsonc
+++ b/knip.jsonc
@@ ... @@
 {
 	"$schema": "https://unpkg.com/knip@latest/schema.json",
-	"entry": ["src/index.ts!", "script/*e2e.js"],
-	"ignoreBinaries": ["gh"],
-	// https://github.com/webpro/knip/issues/417
-	"ignoreDependencies": ["script"],
-	"ignoreExportsUsedInFile": {
-		"interface": true,
-		"type": true
-	},
-	"project": ["src/**/*.ts!", "script/**/*.js"]
+	"entry": ["src/index.ts!"],
+	"ignoreExportsUsedInFile": { "interface": true, "type": true },
+	"project": ["src/**/*.ts!"]
 }"
`;

exports[`expected file changes > package.json 1`] = `
"--- a/package.json
+++ b/package.json
@@ ... @@
 	"scripts": {
 		"build": "tsup",
 		"format": "prettier \\"**/*\\" --ignore-unknown",
-		"initialize": "tsx ./bin/index.js --mode initialize",
 		"lint": "eslint . .*js --max-warnings 0",
 		"lint:knip": "knip",
 		"lint:md": "markdownlint \\"**/*.md\\" \\".github/**/*.md\\" --rules sentences-per-line",
@@ ... @@
 		"prepare": "husky install",
 		"should-semantic-release": "should-semantic-release --verbose",
 		"test": "vitest",
-		"test:create": "node script/create-test-e2e.js",
-		"test:initialize": "node script/initialize-test-e2e.js",
-		"test:migrate": "vitest run -r script/",
 		"tsc": "tsc"
 	},
 	"lint-staged": {
@@ ... @@
 		"vitest": "^1.0.2",
 		"yaml-eslint-parser": "^1.2.2"
 	},
-	"packageManager": "pnpm@8.13.1",
+	"packageManager": "pnpm@8.7.0",
 	"engines": {
 		"node": ">=18"
 	},"
`;
